UnitystationLauncher/Services/InstallationService.cs

```diff
@@ -422,26 +446,53 @@ private async Task StartDownloadAsync(Download download)
                 {
                     ZipArchive archive = new(progressStream);
 
-                    // TODO: Enable extraction cancelling
-                    archive.ExtractToDirectory(download.InstallPath, true);
-
-                    Log.Information("Download completed");
 
-                    _installations.Add(new()
+                    //TODO UI
+                    Action<string> info = new Action<string>((string log) =>
                     {
-                        BuildVersion = download.BuildVersion,
-                        ForkName = download.ForkName,
-                        InstallationId = Guid.NewGuid(),
-                        InstallationPath = download.InstallPath,
-                        LastPlayedDate = DateTime.Now
+                        Console.WriteLine($"info {log}");
+                        InfoList.Add(log);
                     });
+                    Action<string> errors = new Action<string>((string log) =>
+                    {
+                        Console.WriteLine($"error {log}");
+                        ErrorList.Add(log);
+                    });
+                    var scanTask = _codeScanService.OnScan(archive, download.InstallPath, download.GoodFileVersion,
+                        info, errors);
+                    scanTask.Wait();
+                    if (scanTask.Result)
+                    {
+                        Log.Information("Download completed");
+
+                        _installations.Add(new()
+                        {
+                            BuildVersion = download.BuildVersion,
+                            ForkName = download.ForkName,
+                            InstallationId = Guid.NewGuid(),
+                            InstallationPath = download.InstallPath,
+                            LastPlayedDate = DateTime.Now
+                        });
+
+                        WriteInstallations();
+                        EnsureExecutableFlagOnUnixSystems(download.InstallPath);
+                    }
+                    else
+                    {
+                        string jsonString = System.Text.Json.JsonSerializer.Serialize(ErrorList);
+
+                        string directoryPath = AppDomain.CurrentDomain.BaseDirectory;
```

Comment:
You can't write to this directory, this needs to go into the userdata directory.